source/extensions/stat_sinks/metrics_service/grpc_metrics_service_impl.h

```diff
@@ -20,79 +20,106 @@ namespace Extensions {
-  void send(envoy::service::metrics::v3::StreamMetricsMessage& message) override;
+  void send(MetricsPtr&& metrics) override;
 
   // Grpc::AsyncStreamCallbacks
   void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override { stream_ = nullptr; }
 
 private:
-  Grpc::AsyncStream<envoy::service::metrics::v3::StreamMetricsMessage> stream_{};
-  Grpc::AsyncClient<envoy::service::metrics::v3::StreamMetricsMessage,
-                    envoy::service::metrics::v3::StreamMetricsResponse>
-      client_;
   const LocalInfo::LocalInfo& local_info_;
   const Protobuf::MethodDescriptor& service_method_;
   const envoy::config::core::v3::ApiVersion transport_api_version_;
 };
 
 using GrpcMetricsStreamerImplPtr = std::unique_ptr<GrpcMetricsStreamerImpl>;
 
+class MetricsFlusher {
+public:
+  MetricsFlusher(const bool report_counters_as_deltas)
+      : report_counters_as_deltas_(report_counters_as_deltas) {}
+
+  MetricsPtr flush(Stats::MetricSnapshot& snapshot) const;
+
+private:
+  void flushCounter(io::prometheus::client::MetricFamily* metrics_family,
+                    const Stats::MetricSnapshot::CounterSnapshot& counter_snapshot,
+                    int64_t snapshot_time_ms) const;
+  void flushGauge(io::prometheus::client::MetricFamily* metrics_family, const Stats::Gauge& gauge,
+                  int64_t snapshot_time_ms) const;
+  void flushHistogram(io::prometheus::client::MetricFamily* summary_metrics_family,
+                      io::prometheus::client::MetricFamily* histogram_metrics_family,
+                      const Stats::ParentHistogram& envoy_histogram,
+                      int64_t snapshot_time_ms) const;
+
+  const bool report_counters_as_deltas_;
+};
+
 /**
- * Stat Sink implementation of Metrics Service.
+ * Stat Sink that flushes metrics via a gRPC service.
  */
-class MetricsServiceSink : public Stats::Sink {
+template <class RequestProto, class ResponseProto> class MetricsServiceSink : public Stats::Sink {
 public:
   // MetricsService::Sink
-  MetricsServiceSink(const GrpcMetricsStreamerSharedPtr& grpc_metrics_streamer,
-                     const bool report_counters_as_deltas);
-  void flush(Stats::MetricSnapshot& snapshot) override;
-  void onHistogramComplete(const Stats::Histogram&, uint64_t) override {}
+  MetricsServiceSink(
+      const GrpcMetricsStreamerSharedPtr<RequestProto, ResponseProto>& grpc_metrics_streamer,
+      const bool report_counters_as_deltas)
+      : flusher_(report_counters_as_deltas), grpc_metrics_streamer_(grpc_metrics_streamer) {}
 
-  void flushCounter(const Stats::MetricSnapshot::CounterSnapshot& counter_snapshot,
-                    int64_t snapshot_time_ms);
-  void flushGauge(const Stats::Gauge& gauge, int64_t snapshot_time_ms);
-  void flushHistogram(const Stats::ParentHistogram& envoy_histogram, int64_t snapshot_time_ms);
+  void flush(Stats::MetricSnapshot& snapshot) override {
+    grpc_metrics_streamer_->send(std::move(flusher_.flush(snapshot)));
```

Comment:
Is this `std::move` necessary? I think the return value is an rvalue to begin with here